README.md
# 🕷🕸 ПАУК (Spider)
CLI IP Scanner and Port Checker with Telegram Notifications

## 🚧 Disclaimer
The program provided is intended for educational and research purposes only. It may involve techniques that may violate the laws of information security and privacy of private data. The creators and maintainers of this program do not condone or support any illegal activities.

The program should only be used on authorized systems with explicit permission from the owner of the system. The user assumes all responsibility and risk for any actions taken using this program. The creators and maintainers of this program are not responsible for any damages or legal issues that may arise from the misuse of this program.

It is the responsibility of the user to ensure that their actions comply with all applicable laws and regulations regarding information security and privacy. This program should not be used to access or modify any private data without the explicit consent of the owner of that data.

By using this program, you agree to the terms and conditions outlined in this disclaimer.

## Prerequisites
- Python 3.6+
- pip3

## Installation

### Installation for Linux
1. Open a terminal and clone the repository to your local machine.
```
git clone https://github.com/chyornyy/pauk.git
```
2. Navigate to the project directory.
```
cd pauk
```
3. Run the setup script.
```
chmod +x setup.sh && ./setup.sh
```
### Installation for MacOS
1. Open a terminal and clone the repository to your local machine.
```
git clone https://github.com/chyornyy/pauk.git
```
2. Navigate to the project directory.
```
cd pauk
```
3. Run the setup script.
```
chmod +x setup.sh && ./setup.sh
```

### Installation for Windows
Note: The Windows installation option will come soon.

### Creating a Telegram Bot
1. Open the Telegram app.
2. Search for the ["BotFather"](https://t.me/BotFather) bot and start a conversation with it. 
3. Type /newbot and follow the prompts to create a new bot.
4. Once the bot is created, BotFather will provide you with a token. Save this token somewhere secure.
5. Go to your new bot and type /start.

### Getting the Chat ID
1. Open the Telegram app.
2. Search for the ["Get My ID"](https://t.me/getmyid_bot) bot and start a conversation with it.
3. Once you text something it will answer you:
```
Your user ID: <chat_id>
Current chat ID: <chat_id>
```

## Note: It is important to keep your bot token and chat ID secure, as they can be used to access your bot and potentially sensitive information.

## Usage
Activate virtual environment with
```
source venv/bin/activate
```
```
python3 pauk.py <ip_range> --ports <ports> --token <bot_token> --chat <chat_id> -d
```
#### The arguments are described below:
```
<ip_range>: This is a required argument that specifies the IP address range to scan. It should be in CIDR format.
```
```
--ports <ports>: This is an optional argument that allows you to specify a range of ports to scan. The ports can be separated by a hyphen (e.g. 1-1024) or a comma-separated list (e.g. 80,443).
```
```
--token <bot_token>: This is a required argument that specifies the Telegram bot token.
```
```
--chat <chat_id>: This is a required argument that specifies the chat ID of the user or group where you want to receive the notifications.
```
```
-d: This is an optional flag that enables debug mode.
```

### Here are some examples of how to use the program:
#### Example 1: Scan a range of IP addresses for open ports and send the results to a Telegram chat:
Suppose you want to scan a range of IP addresses and send the results to a Telegram chat. You can use this program by providing the IP address range and Telegram bot token and chat ID as arguments. Here is an example command:
```
python3 pauk.py 192.168.0.0/24 -p 80,443 -t <bot_token> -c <chat_id>
```
This command will scan the IP addresses in the range 192.168.0.0/24 on ports 80 and 443 and send the results to the specified Telegram chat.

#### Example 2: Scan a single IP address for open ports:
If you want to scan a single IP address for open ports, you can use this program by providing the IP address and port range as arguments. Here is an example command:
```
python3 pauk.py 192.168.0.1 -p 1-65535 -t <bot_token> -c <chat_id>
```
This command will scan the IP address 192.168.0.1 on all ports between 1 and 65535.

#### Example 3: Scan a range of IP addresses and ports and print the results and debugging info to the console:
If you want to print the results of the port scan to the console, you can use this program by providing the IP address range and port range as arguments. Here is an example command:
```
python3 pauk.py 192.168.0.0/24 -p 1-1024 -t <bot_token> -c <chat_id> -d
```
This command will scan the IP addresses in the range 192.168.0.0/24 on ports from 1 to 1024 and print the results to the console.

## Contributing
Contributions are welcome! Please fork the repository and submit a pull request with your changes.

## License
This project is licensed under the MIT License - see the LICENSE file for details.

## Development Team
#### [Aleksandr Chyornyy](https://github.com/chyornyy) - Backend


CHANGELOG.md
# ПАУК - Changelog

## Version 1.0.0 (2023-04-24)
This is the initial release of the program.

### Added
1. Ping scanning and port scanning functionality.
2. Telegram integration that sends the scan results to a specified chat.
3. IP address range and port range arguments in CIDR and hyphen/comma-separated format respectively.
4. Telegram bot token and chat ID arguments for integration.
5. Debugging option to enable debug information.
6. Progress bar using the tqdm package to display progress during the scan.
7. Retrieval of IP information using the ipinfo.io API.
8. Operating system detection using nmap.

#### The following functions have been added:
- parse_arguments(): Parses command line arguments.
- main(): The main function that performs ping and port scanning, retrieves IP information, and sends the results to a Telegram chat.
- ping(ip_address): Function that pings an IP address to check if it is up.
- scan_ports(ip_address, port_range, args, port_bar): Function that scans ports for an IP address.

#### The following packages have been imported:
- argparse
- ipaddress
- socket
- subprocess
- requests
- tqdm
- aiogram
- nmap3

### Changed
None.
### Deprecated
None.
### Removed
None.
### Fixed
None.
### Security
None.

Note: This program is distributed under the MIT license.

payk.py
from aiogram import Bot, Dispatcher, types
from aiogram.utils import executor
import argparse
import ipaddress
import nmap3
import socket
import subprocess
import requests
import tqdm

nmap = nmap3.Nmap()


def parse_arguments():
    parser = argparse.ArgumentParser(description='Scan a range of IP addresses for open ports and send the results to Telegram.')
    parser.add_argument('ip_range', help='IP address range to scan in CIDR format')
    parser.add_argument('--ports', '-p', default='1-65535', help='Port range to scan, separated by a hyphen (e.g. 1-1024), or a comma-separated list of ports (e.g. 80,443)')
    parser.add_argument('--token', '-t', required=True, help='Telegram bot token')
    parser.add_argument('--chat', '-c', required=True, help='Telegram chat ID')
    parser.add_argument('--debug', '-d', action='store_true', help='Enable debugging information')
    args = parser.parse_args()
    try:
        ip_range = ipaddress.ip_network(args.ip_range)
    except ValueError:
        parser.error('Invalid IP address range')

    # Validate the port range or list of ports
    if '-' in args.ports:
        start_port, end_port = args.ports.split('-')
        if not start_port.isdigit() or not end_port.isdigit() or int(start_port) < 1 or int(end_port) > 65535 or int(start_port) > int(end_port):
            parser.error('Invalid port range')
    else:
        port_list = args.ports.split(',')
        for port in port_list:
            if not port.isdigit() or int(port) < 1 or int(port) > 65535:
                parser.error('Invalid port list')

    return args


def main():
    args = parse_arguments()
    # Clear the results dictionary and the progress bar
    results = {}

    # Parse the IP address and port range
    ip_range = ipaddress.ip_network(args.ip_range)
    if not ip_range:
        print("Please enter a valid IP address range.")
        return

    if '-' in args.ports:
        start_port, end_port = map(int, args.ports.split('-'))
        port_range = range(start_port, end_port + 1)
    else:
        port_range = [int(port) for port in args.ports.split(",")]

    # Set up the Telegram bot
    bot_token = args.token
    chat_id = args.chat
    bot = Bot(token=bot_token)
    dispatcher = Dispatcher(bot)

    # Parse the port range only once
    port_range = list(port_range)

    # Perform ping scanning only if IP address is valid
    try:
        socket.inet_aton(str(ip_range[0]))
    except socket.error:
        if args.debug:
            print(f'Invalid IP address: {ip_range[0]}')
        return

    # Loop through each IP address in the range and perform ping and port scanning
    total_ips = ip_range.num_addresses
    with tqdm.tqdm(total=total_ips, desc='Scanning IPs') as progress_bar:
        for ip_address in ip_range:
            if args.debug:
                print(f'Scanning IP: {ip_address}')

            # Perform ping scanning
            is_up = ping(str(ip_address))
            if not is_up:
                if args.debug:
                    print(f'Host is down: {ip_address}')
                progress_bar.update(1)
                continue

            # If the IP address is live, perform port scanning
            with tqdm.tqdm(total=len(port_range), desc=f'Scanning ports of {ip_address}') as port_bar:
                open_ports = scan_ports(str(ip_address), port_range, args, port_bar)

            # Collect the IP information using ipinfo.io API
            response = requests.get(f"https://ipinfo.io/{ip_address}/json")
            data = response.json()

            # Determine the Operating System
            os_info = get_os_info(str(ip_address))
            message = ""
            if open_ports:
                message += f"IP Address {ip_address} is live.\nOpen ports: {open_ports}\n\n"
                message += f"City: {data['city']}\nRegion: {data['region']}\nCountry: {data['country']}\n"
                message += f"Timezone: {data['timezone']}\n"
                if 'postal' in data:
                    message += f"Postal: {data['postal']}\n\n"
                else:
                    message += "Postal: Unknown\n\n"
                if 'hostname' in data:
                    message += f"Hostname: {data['hostname']}\n"
                else:
                    message += "Hostname: Unknown\n"
                message += f"Organization: {data['org']}\n\n"
                message += f"OS: {os_info}"

                if args.debug:
                    print(message)
                    try:
                        async def send_message():
                            await bot.send_message(chat_id=chat_id, text=message)
                        executor.start(dispatcher, send_message())
                    except Exception as e:
                        if args.debug:
                            print(f"Error sending message to Telegram: {str(e)}")
                print("Scan complete.")


def ping(ip_address):
    ping_cmd = ['ping', '-c', '1', '-W', '1', ip_address]
    ping_response = subprocess.run(ping_cmd, capture_output=True)
    return ping_response.returncode == 0


def scan_ports(ip_address, port_range, args, port_bar):
    open_ports = []
    for port in port_range:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.settimeout(1)
            try:
                s.connect((ip_address, port))
                open_ports.append(port)
            except (socket.timeout, ConnectionRefusedError):
                pass
            except Exception as e:
                if args.debug:
                    print(f"Error scanning port {port} of {ip_address}: {e}")
        port_bar.update(1)
    return open_ports


def get_os_info(ip_address):
    default_value = 'Unknown'
    try:
        result = nmap.nmap_os_detection(ip_address)
        if ip_address in result:
            os_matches = result[ip_address].get('osmatch', [])
            if os_matches:
                os_info = os_matches[0]['name']
            else:
                os_info = default_value
        else:
            os_info = default_value
    except nmap3.exceptions.NmapExecutionFailed:
        os_info = default_value
    return os_info


if __name__ == '__main__':
    main()


pauk.py
from aiogram import Bot, Dispatcher, types
from aiogram.utils import executor
import argparse
import ipaddress
import nmap3
import requests
import time
import socket
import subprocess
import tqdm

import functions as fn

nmap = nmap3.Nmap()

def parse_arguments():
    parser = argparse.ArgumentParser(description='Scan a range of IP addresses for open ports and send the results to Telegram.')
    parser.add_argument('ip_range', help='IP address range to scan in CIDR format')
    parser.add_argument('--ports', '-p', default='1-65535', help='Port range to scan, separated by a hyphen (e.g. 1-1024), or a comma-separated list of ports (e.g. 80,443)')
    parser.add_argument('--token', '-t', required=True, help='Telegram bot token')
    parser.add_argument('--chat', '-c', required=True, help='Telegram chat ID')
    parser.add_argument('--debug', '-d', action='store_true', help='Enable debugging information')
    args = parser.parse_args()
    try:
        ip_range = ipaddress.ip_network(args.ip_range)
    except ValueError:
        parser.error('Invalid IP address range')

    # Validate the port range or list of ports
    if '-' in args.ports:
        start_port, end_port = args.ports.split('-')
        if not start_port.isdigit() or not end_port.isdigit() or int(start_port) < 1 or int(end_port) > 65535 or int(start_port) > int(end_port):
            parser.error('Invalid port range')
    else:
        port_list = args.ports.split(',')
        for port in port_list:
            if not port.isdigit() or int(port) < 1 or int(port) > 65535:
                parser.error('Invalid port list')

    return args


def main():
    args = parse_arguments()
    # Clear the results dictionary and the progress bar
    results = {}

    # Parse the IP address and port range
    ip_range = ipaddress.ip_network(args.ip_range)
    if not ip_range:
        print("Please enter a valid IP address range.")
        return

    if '-' in args.ports:
        start_port, end_port = map(int, args.ports.split('-'))
        port_range = range(start_port, end_port + 1)
    else:
        port_range = [int(port) for port in args.ports.split(",")]

    # Set up the Telegram bot
    bot_token = args.token
    chat_id = args.chat
    bot = Bot(token=bot_token)
    dispatcher = Dispatcher(bot)

    # Parse the port range only once
    port_range = list(port_range)

    # Perform ping scanning only if IP address is valid
    try:
        socket.inet_aton(str(ip_range[0]))
    except socket.error:
        if args.debug:
            print(f'Invalid IP address: {ip_range[0]}')
        return

    # Loop through each IP address in the range and perform ping and port scanning
    total_ips = ip_range.num_addresses
    with tqdm.tqdm(total=total_ips, desc='Scanning IPs') as progress_bar:
        start_time = time.time()
        alive_hosts_amount = 0
        checked_hosts_amount = 0
        for ip_address in ip_range:
            checked_hosts_amount += 1
            if args.debug:
                print(f'Scanning IP: {ip_address}')

            # Perform ping scanning
            is_up = ping(str(ip_address))
            if not is_up:
                if args.debug:
                    print(f'Host is down: {ip_address}')
                progress_bar.update(1)
                continue

            # If the IP address is live, perform port scanning
            with tqdm.tqdm(total=len(port_range), desc=f'Scanning ports of {ip_address}') as port_bar:
                alive_hosts_amount += 1
                open_ports = scan_ports(str(ip_address), port_range, args, port_bar)
            
            # Collect the IP information using ipinfo.io API
            response = requests.get(f"https://ipinfo.io/{ip_address}/json")
            data = response.json()

            # Determine the Operating System
            os_info = get_os_info(str(ip_address))
            message = ""
            if open_ports:
                message += f"IP Address {ip_address} is live.\nOpen ports: {open_ports}\n\n"
                message += f"City: {data['city']}\nRegion: {data['region']}\nCountry: {data['country']}\n"
                message += f"Timezone: {data['timezone']}\n"
                if 'postal' in data:
                    message += f"Postal: {data['postal']}\n\n"
                else:
                    message += "Postal: Unknown\n\n"
                if 'hostname' in data:
                    message += f"Hostname: {data['hostname']}\n"
                else:
                    message += "Hostname: Unknown\n"
                message += f"Organization: {data['org']}\n\n"
                message += f"OS: {os_info}"

                if args.debug:
                    print(message)
                    try:
                        async def send_message():
                            await bot.send_message(chat_id=chat_id, text=message)
                        executor.start(dispatcher, send_message())
                    except Exception as e:
                        if args.debug:
                            print(f"Error sending message to Telegram: {str(e)}")
                end_time = time.time()
                time_taken = int(end_time - start_time)
                print(f"Scan complete!\nScanning Time: {fn.format_time(time_taken)}.\nTotal hosts scanned: {checked_hosts_amount}.\nFound hosts alive: {alive_hosts_amount}.")


def ping(ip_address):
    ping_cmd = ['ping', '-c', '1', '-W', '1', ip_address]
    ping_response = subprocess.run(ping_cmd, capture_output=True)
    return ping_response.returncode == 0


def scan_ports(ip_address, port_range, args, port_bar):
    open_ports = []
    for port in port_range:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.settimeout(1)
            try:
                s.connect((ip_address, port))
                open_ports.append(port)
            except (socket.timeout, ConnectionRefusedError):
                pass
            except Exception as e:
                if args.debug:
                    print(f"Error scanning port {port} of {ip_address}: {e}")
        port_bar.update(1)
    return open_ports


def get_os_info(ip_address):
    default_value = 'Unknown'
    try:
        result = nmap.nmap_os_detection(ip_address)
        if ip_address in result:
            os_matches = result[ip_address].get('osmatch', [])
            if os_matches:
                os_info = os_matches[0]['name']
            else:
                os_info = default_value
        else:
            os_info = default_value
    except nmap3.exceptions.NmapExecutionFailed:
        os_info = default_value
    return os_info


if __name__ == '__main__':
    main()
